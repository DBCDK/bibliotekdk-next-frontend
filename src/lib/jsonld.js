import { getCanonicalWorkUrl } from "./utils";

/**
 * Maps materialType to schema.org BookFormatType
 *
 * @param {string} materialType
 *
 * @returns {string} schema.org BookFormatType
 */
function getSchemaOrgBookFormat(materialType) {
  /** https://schema.org/BookFormatType
   * AudiobookFormat
   * EBook
   * GraphicNovel
   * Hardcover
   * Paperback
   */

  materialType = materialType.toLowerCase();

  if (materialType.includes("ebog")) {
    return "http://schema.org/EBook";
  } else if (materialType.includes("lydbog")) {
    return "http://schema.org/AudiobookFormat";
  } else if (materialType.includes("graphic")) {
    return "http://schema.org/GraphicNovel";
  } else if (materialType.includes("bog")) {
    // Hardcover or paperback?
    return "http://schema.org/Paperback";
  }
  return null;
}

/**
 * Part real data and part mocked data
 * We still need some more data to be available through the API
 * - https://developers.google.com/search/docs/data-types/book
 * - https://solsort.dk/dkabm-til-schema.org
 *
 *
 * Still needs a lot of work, we need more data to be available via API
 *
 * @param {object} work
 *
 * @returns {object} json ld representation of work
 */
export function getJSONLD({
  id,
  title,
  creators = [],
  path,
  materialTypes = [],
}) {
  const res = {
    "@context": "https://schema.org",
    "@type": "WebPage",
    breadcrumb: path.join(" > "),
    mainEntity: {
      "@id": id,
      "@type": "Book",
      author: creators.map((creator) => ({
        "@type": "Person",
        name: creator.name,
      })),
      name: title,

      url: getCanonicalWorkUrl({ title, creators, id }),
      workExample: materialTypes.map((entry) => {
        // edition/manifestation
        const manifestation = {
          "@id": entry.pid,
          "@type": "Book",
          author: creators.map((creator) => ({
            "@type": "Person",
            name: creator.name,
          })),
          //   bookEdition: '' // TODO
          datePublished: "2015-05-01", // TODO
          //   identifier // TODO
          // name: '' // only if its different, TODO
          //   sameAs: '' // TODO
          inLanguage: "Dansk", // TODO
          isbn: "00000000", // TODO
          publisher: { "@type": "Organization", name: "Some publisher" }, // TODO
          //   url // TODO
          //   potentialAction: {} // TODO
        };
        if (entry.cover) {
          manifestation.image = entry.cover.detail;
        }
        const bookFormat = getSchemaOrgBookFormat(entry.materialType);
        if (bookFormat) {
          manifestation.bookFormat = bookFormat;
        }
        return manifestation;
      }),
      //   sameAs: '' // recommended
    },
  };
  return res;
}
