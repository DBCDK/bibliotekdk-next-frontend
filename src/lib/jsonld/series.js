/**
 * Creates JSON-LD representation of the article
 * - https://developers.google.com/search/docs/data-types/article
 *
 *
 * @param {Object} series
 *
 * @returns {Object} JSON-LD representation of series
 */
export function getJSONLD({ series }) {
  if (!series) {
    return null;
  }
  const url = `https://bibliotek.dk/serie/${series.seriesId}`;
  const res = {
    "@context": "https://schema.org",
    "@type": "CreativeWorkSeries",
    mainEntityOfPage: {
      "@type": "WebPage",
      "@id": url,
    },
    name: series.title,
    description: series.description,
    url: url,
    genre: series.seriesWorkTypes,
    inLanguage: series.mainLanguages?.join(", "),
    hasPart: series.members.map((member) => ({
      "@type": "CreativeWork",
      name: member.work.titles.main[0],
      url: `https://bibliotek.dk/work/${member.work.workId}`,
      description: member.work.abstract?.[0],
      genre: member.work.workTypes?.join(", "),
      creator: member.work.creators.map((creator) => ({
        "@type": "Person",
        name: creator.name,
      })),
      image: {
        "@type": "ImageObject",
        url: member.work.manifestations.mostRelevant[0]?.cover?.detail,
        thumbnail: member.work.manifestations.mostRelevant[0]?.cover?.thumbnail,
      },
    })),
    author: {
      "@type": "Organization",
      name: "Bibliotek.dk",
    },
    publisher: {
      "@type": "Organization",
      name: "Bibliotek.dk",
      logo: {
        "@type": "ImageObject",
        url: "https://bibliotek.dk/img/logo.png",
      },
    },
  };
  // find an image to represent the serie
  const serieImage = res?.hasPart?.find((part) => part?.image?.url);
  res.image = serieImage && serieImage?.image?.url;

  return res;
}
