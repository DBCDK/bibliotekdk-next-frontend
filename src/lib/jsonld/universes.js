/**
 * Creates JSON-LD representation of the article
 * - https://developers.google.com/search/docs/data-types/article
 *
 *
 * @param {Object} universe
 *
 * @returns {Object} JSON-LD representation of article
 */
export function getJSONLD({ universe }) {
  const url = `https://bibliotek.dk/univers/${universe.universeId}`;
  const res = {
    "@context": "https://schema.org",
    "@type": "CreativeWork",
    mainEntityOfPage: {
      "@type": "WebPage",
      "@id": url,
    },
    name: universe.title,
    description: universe.description,
    url: url,
    genre: universe.workTypes,
    author: {
      "@type": "Organization",
      name: "Bibliotek.dk",
    },
    publisher: {
      "@type": "Organization",
      name: "Bibliotek.dk",
      logo: {
        "@type": "ImageObject",
        url: "https://bibliotek.dk/img/logo.png",
      },
    },
  };
  return res;
}
