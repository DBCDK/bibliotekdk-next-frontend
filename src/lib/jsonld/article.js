import { getCanonicalArticleUrl } from "@/lib/utils";

/**
 * Creates JSON-LD representation of the article
 * - https://developers.google.com/search/docs/data-types/article
 *
 *
 * @param {object} article
 *
 * @returns {object} JSON-LD representation of article
 */
export function getJSONLD({
  nid,
  title,
  fieldRubrik,
  fieldImage,
  body,
  entityCreated,
  entityModified,
}) {
  const url = getCanonicalArticleUrl({ title, nid });
  const res = {
    "@context": "https://schema.org",
    "@type": "Article",
    mainEntityOfPage: {
      "@type": "WebPage",
      "@id": url,
    },
    headline: title,
    image: [fieldImage],
    datePublished: entityCreated,
    dateModified: entityModified,
    description: fieldRubrik,
    articleBody: body && body.value,
    author: {
      "@type": "Organization",
      name: "Bibliotek.dk",
    },
    publisher: {
      "@type": "Organization",
      name: "Bibliotek.dk",
      logo: {
        "@type": "ImageObject",
        url: "https://beta.bibliotek.dk/img/logo.png",
      },
    },
  };
  return res;
}
