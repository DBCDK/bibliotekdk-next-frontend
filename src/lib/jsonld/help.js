import { getCanonicalArticleUrl } from "@/lib/utils";
import Translate from "@/components/base/translate";

/**
 * Creates JSON-LD representation of the (help) article
 * - https://developers.google.com/search/docs/data-types/article
 *
 *
 * @param {object} article
 *
 * @returns {object} JSON-LD representation of article
 */
export function getJSONLD({
  nid,
  title,
  fieldImage,
  body,
  entityCreated,
  entityChanged,
}) {
  const url = getCanonicalArticleUrl({ title, nid });
  const description = Translate({
    context: "metadata",
    label: "help-description",
  });
  const res = {
    "@context": "https://schema.org",
    "@type": "Article",
    mainEntityOfPage: {
      "@type": "WebPage",
      "@id": url,
    },
    headline: title,
    image: [fieldImage],
    datePublished: entityCreated,
    dateModified: entityChanged,
    description,
    articleBody: body && body.value,
    author: {
      "@type": "Organization",
      name: "Bibliotek.dk",
    },
    publisher: {
      "@type": "Organization",
      name: "Bibliotek.dk",
      logo: {
        "@type": "ImageObject",
        url: "https://beta.bibliotek.dk/img/logo.png",
      },
    },
  };

  return res;
}
