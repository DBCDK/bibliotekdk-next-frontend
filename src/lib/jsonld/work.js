import { getCanonicalWorkUrl } from "@/lib/utils";

/**
 * Maps materialType to schema.org BookFormatType
 *
 * @param {string} materialType
 *
 * @returns {string} schema.org BookFormatType
 */
function getSchemaOrgBookFormat(materialType) {
  /** https://schema.org/BookFormatType
   * AudiobookFormat
   * EBook
   * GraphicNovel
   * Hardcover
   * Paperback
   */

  materialType = materialType.toLowerCase();

  if (materialType.includes("e-bog")) {
    return "http://schema.org/EBook";
  } else if (materialType.includes("lydbog")) {
    return "http://schema.org/AudiobookFormat";
  } else if (materialType.includes("graphic")) {
    return "http://schema.org/GraphicNovel";
  } else if (materialType.includes("bog")) {
    // Hardcover or paperback?
    return "http://schema.org/Paperback";
  }
  return null;
}

/**
 * Creates JSON-LD representation ofthe work
 * - https://developers.google.com/search/docs/data-types/book
 * - https://solsort.dk/dkabm-til-schema.org
 *
 *
 * Note:
 *  - It is recommended for work and manifestations to have "sameAs" field,
 *    pointing to reference web page.
 *  - "potentialAction" is not implemented
 *  -
 *
 * @param {object} work
 *
 * @returns {object} JSON-LD representation of work
 */
export function getJSONLD({
  id,
  title,
  description,
  creators = [],
  path,
  manifestations = [],
}) {
  const url = getCanonicalWorkUrl({ title, creators, id });
  const res = {
    "@context": "https://schema.org",
    "@type": "WebPage",
    breadcrumb: path.join(" > "),
    mainEntity: {
      "@id": id,
      "@type": "Book",
      abstract: description,
      author: creators.map((creator) => ({
        "@type": "Person",
        name: creator.name,
      })),
      name: title,
      url,
      workExample: manifestations.map((entry) => {
        const manifestation = {
          "@id": entry.pid,
          "@type": "Book",
          author: entry.creators.map((creator) => ({
            "@type": "Person",
            name: creator.name,
          })),
          datePublished: entry.datePublished,
          identifier: entry.pid,
          url,
          inLanguage: entry.inLanguage,
        };
        if (entry.title && title !== entry.title) {
          // only add name to manifestation if it differs from work
          manifestation.name = entry.title;
        }
        if (entry.isbn) {
          manifestation.isbn = entry.isbn;
        }
        if (entry.cover) {
          manifestation.image = entry.cover.detail;
        }
        if (entry.edition) {
          manifestation.bookEdition = entry.edition;
        }
        if (entry.publisher) {
          manifestation.publisher = {
            "@type": "Organization",
            name: entry.publisher,
          };
        }
        const bookFormat = getSchemaOrgBookFormat(entry.materialType);
        if (bookFormat) {
          manifestation.bookFormat = bookFormat;
        }
        return manifestation;
      }),
    },
  };
  return res;
}
