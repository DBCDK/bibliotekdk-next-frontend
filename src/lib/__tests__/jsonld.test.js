import { getJSONLD } from "../jsonld";
import dummy_workDataApi from "../../components/work/dummy.workDataApi";

test("generateJSONLD", () => {
  const data = dummy_workDataApi({ workId: "some-id" });
  const actual = getJSONLD({ ...data.work, id: "some-id" });
  expect(actual).toMatchSnapshot();
});
