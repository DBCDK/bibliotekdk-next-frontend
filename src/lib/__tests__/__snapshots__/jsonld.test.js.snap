// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateJSONLD 1`] = `
Object {
  "@context": "https://schema.org",
  "@type": "WebPage",
  "breadcrumb": "Bøger > Fiktion > skønlitteratur > roman",
  "mainEntity": Object {
    "@id": "some-id",
    "@type": "Book",
    "abstract": "some description",
    "author": Array [
      Object {
        "@type": "Person",
        "name": "h g wells",
      },
    ],
    "name": "Klodernes kamp",
    "url": "https://bibliotek.dk/klodernes-kamp_h-g-wells/some-id",
    "workExample": Array [
      Object {
        "@id": "870970-basis:06442870",
        "@type": "Book",
        "author": Array [
          Object {
            "@type": "Person",
            "name": "some contributor",
          },
        ],
        "bookFormat": "http://schema.org/Paperback",
        "datePublished": "2020",
        "identifier": "870970-basis:06442870",
        "image": null,
        "inLanguage": "da",
        "url": "https://bibliotek.dk/klodernes-kamp_h-g-wells/some-id",
      },
    ],
  },
}
`;
